### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry): Define affine morphisms

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/Morphisms/Affine.lean [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+/-- A morphism of schemes `X ⟶ Y` is affine if the preimages of affine open sets are affine. -/
+@[mk_iff]
+class IsAffineHom {X Y : Scheme} (f : X ⟶ Y) : Prop where
+  isAffine_preimage : ∀ U : Opens Y, IsAffineOpen U → IsAffineOpen (f ⁻¹ᵁ U)
+
+/-- The preimage of an affine open as an `Scheme.affine_opens`. -/
+@[simps]
+def affinePreimage {X Y : Scheme} (f : X ⟶ Y) [IsAffineHom f] (U : Y.affineOpens) :
+    X.affineOpens :=
+⟨f ⁻¹ᵁ U.1, IsAffineHom.isAffine_preimage _ U.prop⟩
```
Comment (joelriou) [inline Mathlib/AlgebraicGeometry/Morphisms/Affine.lean:34]:
```suggestion
  ⟨f ⁻¹ᵁ U.1, IsAffineHom.isAffine_preimage _ U.prop⟩
```

File: Mathlib/AlgebraicGeometry/Morphisms/Affine.lean [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+instance (priority := 900) [IsIso f] : IsAffineHom f :=
+⟨fun _ hU ↦ hU.map_isIso f⟩
```
Comment (joelriou) [inline Mathlib/AlgebraicGeometry/Morphisms/Affine.lean:37]:
```suggestion
  ⟨fun _ hU ↦ hU.map_isIso f⟩
```

File: Mathlib/AlgebraicGeometry/Morphisms/Affine.lean [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+instance (priority := 900) [IsAffineHom f] : QuasiCompact f :=
+(quasiCompact_iff_forall_affine f).mpr (fun U hU ↦ (IsAffineHom.isAffine_preimage U hU).isCompact)
```
Comment (joelriou) [inline Mathlib/AlgebraicGeometry/Morphisms/Affine.lean:40]:
```suggestion
  (quasiCompact_iff_forall_affine f).mpr
    (fun U hU ↦ (IsAffineHom.isAffine_preimage U hU).isCompact)
```

File: Mathlib/AlgebraicGeometry/Morphisms/Affine.lean [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+  apply IsAffineHom.isAffine_preimage
+  exact hU
+
+instance : MorphismProperty.IsMultiplicative @IsAffineHom where
+  id_mem := inferInstance
+  comp_mem _ _ _ _ := inferInstance
+
+/-- The `AffineTargetMorphismProperty` corresponding to affine morphisms. -/
+def IsAffineHom.affineProperty : AffineTargetMorphismProperty :=
+fun X _ _ _  ↦ IsAffine X
```
Comment (joelriou) [inline Mathlib/AlgebraicGeometry/Morphisms/Affine.lean:56]:
```suggestion
  fun X _ _ _  ↦ IsAffine X
```

File: Mathlib/AlgebraicGeometry/Morphisms/Affine.lean [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+@[simp] lemma IsAffineHom.affineProperty_toProperty :
+  AffineTargetMorphismProperty.toProperty IsAffineHom.affineProperty f ↔
+    IsAffine Y ∧ IsAffine X := by
```
Comment (joelriou) [inline Mathlib/AlgebraicGeometry/Morphisms/Affine.lean:60]:
```suggestion
    AffineTargetMorphismProperty.toProperty IsAffineHom.affineProperty f ↔
      IsAffine Y ∧ IsAffine X := by
```

File: Mathlib/AlgebraicGeometry/Morphisms/Affine.lean [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  delta AffineTargetMorphismProperty.toProperty IsAffineHom.affineProperty; simp
+
+lemma isAffineHom_iff_affineProperty :
+    IsAffineHom f ↔ targetAffineLocally IsAffineHom.affineProperty f :=
+(isAffineHom_iff f).trans ⟨fun H U ↦ H U U.prop, fun H U hU ↦ H ⟨U, hU⟩⟩
```
Comment (joelriou) [inline Mathlib/AlgebraicGeometry/Morphisms/Affine.lean:65]:
```suggestion
  (isAffineHom_iff f).trans ⟨fun H U ↦ H U U.prop, fun H U hU ↦ H ⟨U, hU⟩⟩
```

File: Mathlib/AlgebraicGeometry/Morphisms/Affine.lean [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+Authors: Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.Morphisms.QuasiSeparated
+import Mathlib.AlgebraicGeometry.Morphisms.OpenImmersion
+
+/-!
+
+# Affine morphisms of schemes
+
+A morphism of schemes `f : X ⟶ Y` is affine if the preimage of affine opens are affine.
```
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/Morphisms/Affine.lean:13]:
```suggestion
A morphism of schemes `f : X ⟶ Y` is affine if the preimage of an arbitrary affine open subset of `Y` is affine.

It is equivalent to ask only that `Y` is covered by affine opens whose preimage is affine.

## Main theorems 

(local on target, stable under base change)
```

File: Mathlib/AlgebraicGeometry/Morphisms/Affine.lean [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+universe v u
+
+open CategoryTheory TopologicalSpace Opposite
+
+namespace AlgebraicGeometry
+
+variable {X Y Z : Scheme.{u}} (f : X ⟶ Y) (g : Y ⟶ Z)
+
+/-- A morphism of schemes `X ⟶ Y` is affine if the preimages of affine open sets are affine. -/
```
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/Morphisms/Affine.lean:25]:
```suggestion
/-- A morphism of schemes `X ⟶ Y` is affine if the preimage of any affine open subset of `Y` is affine. -/
```

**Review by kbuzzard — COMMENTED** [review]
This looks great but it would be even better if we had a more detailed module docstring.

File: Mathlib/AlgebraicGeometry/Morphisms/Affine.lean [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+  intros U hU
+  fapply (Scheme.Hom.isAffineOpen_iff_of_isOpenImmersion (Scheme.ιOpens _)).mp
+  convert hU.basicOpen (X.presheaf.map (homOfLE le_top).op r)
+  rw [X.basicOpen_res]
+  ext1
+  refine Set.image_preimage_eq_inter_range.trans ?_
+  erw [Subtype.range_coe]
+  rfl
+
+lemma iSup_basicOpen_eq_top_of_span_eq_top {X : Scheme} (s : Set (X.presheaf.obj (op ⊤)))
```
Comment (joelriou) [inline Mathlib/AlgebraicGeometry/Morphisms/Affine.lean:101]: This PR looks very good to me, apart from the fact that this lemma (and the next) may be moved to a slightly better file.

**Comment by joelriou** [comment]
Thanks!

bors merg

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**